hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java

```diff
@@ -431,55 +477,128 @@ public OwnerOfAndTokenURIWrapper decodeTokenUriNFT(final Bytes input) {
-        final var serialNo = (BigInteger) decodedArguments.get(0);
-        return new GetApprovedWrapper(serialNo.longValue());
+                        input,
+                        offset == 0
+                                ? ERC_GET_APPROVED_FUNCTION_SELECTOR
+                                : HAPI_GET_APPROVED_FUNCTION_SELECTOR,
+                        offset == 0
+                                ? ERC_GET_APPROVED_FUNCTION_DECODER
+                                : HAPI_GET_APPROVED_FUNCTION_DECODER);
+
+        final var tokenId =
+                offset == 0
+                        ? impliedTokenId
+                        : convertAddressBytesToTokenID(decodedArguments.get(0));
+
+        final var serialNo = (BigInteger) decodedArguments.get(offset);
+        return new GetApprovedWrapper(tokenId, serialNo.longValue());
     }
 
     public TokenAllowanceWrapper decodeTokenAllowance(
-            final Bytes input, final UnaryOperator<byte[]> aliasResolver) {
+            final Bytes input,
+            final TokenID impliedTokenId,
+            final UnaryOperator<byte[]> aliasResolver) {
+        final var offset = impliedTokenId == null ? 1 : 0;
+
         final Tuple decodedArguments =
-                decodeFunctionCall(input, TOKEN_ALLOWANCE_SELECTOR, TOKEN_ALLOWANCE_DECODER);
+                decodeFunctionCall(
+                        input,
+                        offset == 0 ? ERC_ALLOWANCE_SELECTOR : HAPI_ALLOWANCE_SELECTOR,
+                        offset == 0 ? ERC_ALLOWANCE_DECODER : HAPI_ALLOWANCE_DECODER);
 
+        final var tokenId =
+                offset == 0
+                        ? impliedTokenId
+                        : convertAddressBytesToTokenID(decodedArguments.get(0));
         final var owner =
-                convertLeftPaddedAddressToAccountId(decodedArguments.get(0), aliasResolver);
+                convertLeftPaddedAddressToAccountId(decodedArguments.get(offset), aliasResolver);
         final var spender =
-                convertLeftPaddedAddressToAccountId(decodedArguments.get(1), aliasResolver);
+                convertLeftPaddedAddressToAccountId(
+                        decodedArguments.get(offset + 1), aliasResolver);
 
-        return new TokenAllowanceWrapper(owner, spender);
+        return new TokenAllowanceWrapper(tokenId, owner, spender);
     }
 
     public ApproveWrapper decodeTokenApprove(
             final Bytes input,
-            final TokenID token,
+            final TokenID impliedTokenId,
             final boolean isFungible,
-            final UnaryOperator<byte[]> aliasResolver) {
-        final Tuple decodedArguments =
-                decodeFunctionCall(input, TOKEN_APPROVE_SELECTOR, TOKEN_APPROVE_DECODER);
+            final UnaryOperator<byte[]> aliasResolver,
+            WorldLedgers ledgers) {
+
+        final var offset = impliedTokenId == null ? 1 : 0;
+        Bytes selector;
+        if (offset == 0) {
```

Comment:
```suggestion
        final var tokenIdArgumentIndex = impliedTokenId == null ? 1 : 0;
        final var offset = impliedTokenId != null;
        if (offset) {
```

Reply:
Could use this logic in most of the  added functions.

Reply:
then we get sonar complaints about the magic constant 1.  JVM will compile that and do a similar conversion for us at runtime.

I think I should just go back to the nested ternary and ignore sonar.  It was perfectly readable.

Reply:
I also need the integer value of offset at 562 and a few other places later.

Reply:
Got it. As is is fine then